File: concepts/impl/binary_search.go
package main

import "fmt"

func binary(x int, data []int) []int {

	mid := len(data) / 2
	if len(data) > 1 {

		if data[mid] < x {
			return data[mid:len(data)]
		}

		return data[0:mid]
	}

	return data
}

func main() {
	data := []int{1, 24, 34, 454, 3434, 34344}
	x := 34344

	found := binary(x, data)
	for len(data) >= 2 {
		found = binary(x, found)
		fmt.Println(found)
	}

}

File: impl/c/binary_search.c
#include  <stdio.h>

// returns 0 of if it found the value else 1 as default
int bsearch(int tofind, int *arr, int len) {
  if (arr[0] == tofind || arr[1] == tofind) return 0;

  int head = 0;
  int tail = len - 1;
  int mid = len / 2;

  while (head <= tail) {
    int peak = arr[mid];
    if (peak == tofind) return 0;
    if(peak > tofind) {
      tail = mid -1;
      mid = (head + tail) /2;
    }
    head = mid +1;
    mid = (head+tail)/2;
  }
  return 1;
}

int main(void) {
  int arr[] = {1, 2, 3, 4, 5, 6, 7, 8};
  int search0 = bsearch(11, arr, sizeof(arr)/sizeof(int));
  int search1 = bsearch(10, arr, sizeof(arr)/sizeof(int));

  printf("%d\n", search0);
  printf("%d\n", search1);
}

